### PR TITLE
Propagate gcs-connector options to GcsUtil

### PR DESCRIPTION
**Pull request by clairemcginty**
Context: I was reading GCS Parquet files via SplittableDoFn and noticed that `ReadableFile#openSeekable` does not propagate any of the [gcs-connector options](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/v2.2.25/gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java#L66-L562) specified in my `core-site.xml` file. Particularly, I wanted to turn off [fs.gs.inputstream.fast.fail.on.not.found.enable](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/v2.2.25/gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java#L330), which is redundant in a SDF with default empty-match treatment, and tweak [fs.gs.inputstream.fadvise](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/v2.2.25/gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java#L350). It looks like these `GoogleCloudStorageReadOptions` options need to be set explicitly in GcsUtil, and passed to any GoogleCloudStorage#open calls (see [reference](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/v2.2.25/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java#L802-L846)).

The big downside of this PR is of course, pulling in Hadoop :( The alternative is to manually copy-paste all the Configuration keys manually into `GcsUtil`, which seems harder to maintain. Or, I could omit the `GcsReadOptionsFactory` factory logic entirely and leave it 100% up to the user to construct `GoogleCloudStorageReadOptions` instances.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java
Comment (clairemcginty) [inline sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java:174]: Copy-pasted from here: https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/v2.2.25/gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java#L656-L677

I think we could make a case to make that method public in a future release so we're not pulling in Hadoop.

Or, I could omit this branch entirely and always return `GoogleCloudStorageReadOptions.DEFAULT`, and leave it up to the user to supply a `GoogleCloudStorageReadOptions` instance (thus passing the Hadoop dependency down to the user-end).

File: sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/GcpCoreApiSurfaceTest.java
Comment (clairemcginty) [inline sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/GcpCoreApiSurfaceTest.java:59]: this is unfortunate 😓  GoogleCloudStorageReadOptions has one property that's strictly an ImmutableSet: https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/v2.2.25/gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java#L519

**Comment by clairemcginty** [comment]
assign set of reviewers

**Comment by clairemcginty** [comment]
assigned reviewers since at first glance, the GHA failures did not look related/might be transient? could be wrong though.

**Comment by clairemcginty** [comment]
bumping this PR -- cc @scwhittle since I saw you recently made changes to GcsUtil?

File: sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java
Comment (scwhittle) [reply sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java:174]: > Or, I could omit this if/else branch entirely and always return GoogleCloudStorageReadOptions.DEFAULT, and leave it up to the user to supply a GoogleCloudStorageReadOptions instance (thus passing the Hadoop dependency down to the user-end).

I think this would be preferable to avoid having to pull in the other packages but asking for some other Beam maintainers more familiar with dep management etc to take a look as well.

File: sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java
Comment (scwhittle) [inline sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java:323]: This seems like a bug in `GoogleCloudStorageImpl.open` method that it doesn't use these options but you can remove the separate member variable and use `googleCloudStorageOptions.getReadChannelOptions()` in `open` below below.

Comment (clairemcginty) [reply sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java:323]: yeah `GoogleCloudStorageImpl` set up in a slightly strange way -- it's constructed with a GoogleCloudStorageOptions arg but also accepts a separate GoogleCloudStorageOptions as an argument to `open`, while ignoring the former instance variable. Dropping the separate member variable here makes sense, will do 👍

**Comment by scwhittle** [comment]
R: @shunping (XQ suggested you to help review this)

**Comment by shunping** [comment]
> R: @shunping (XQ suggested you to help review this)

ack. will take a look today

**Comment by clairemcginty** [comment]
@shunping , whenever you have a chance I'd appreciate any feedback on this!

File: sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java
Comment (shunping) [reply sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java:174]: Agree with @scwhittle. It is also easier for maintainers and if new options are added, we don't need to change the code to support that.

File: sdks/java/extensions/google-cloud-platform-core/src/test/resources/test-hadoop-conf.xml
Comment (shunping) [inline sdks/java/extensions/google-cloud-platform-core/src/test/resources/test-hadoop-conf.xml:1]: Could you add license information here? Example: https://github.com/apache/beam/blob/e1e093455acaa9a6d5b131cf5103165a661c7615/sdks/java/build-tools/src/main/resources/beam/checkstyle/checkstyle.xml#L2

**Review by shunping — CHANGES_REQUESTED** [review]
Thanks for adding this option. I left some minor comments there, and overall it looks good. 

After you finish revising, could you please sync to the current HEAD and so the tests are re-triggered? Thanks!

File: sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java
Comment (clairemcginty) [reply sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtil.java]: updated!

File: sdks/java/extensions/google-cloud-platform-core/src/test/resources/test-hadoop-conf.xml
Comment (clairemcginty) [reply sdks/java/extensions/google-cloud-platform-core/src/test/resources/test-hadoop-conf.xml]: Ended up removing the file

**Comment by shunping** [comment]
Run Java PreCommit

**Comment by shunping** [comment]
Running the failed precommit test again, though the failure seems unrelated to the code change here.

File: sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/GcpCoreApiSurfaceTest.java
Comment (shunping) [inline sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/GcpCoreApiSurfaceTest.java:58]: Looks like this break a test:
https://ge.apache.org/s/5ubkbvlvrjzgi/console-log/task/:sdks:java:extensions:google-cloud-platform-core:test?anchor=5&page=1

Comment (clairemcginty) [reply sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/GcpCoreApiSurfaceTest.java:58]: thanks for tracking that down! Pushed a fix. one of the precommit tests is still failing though:

```
org.apache.beam.runners.dataflow.worker.streaming.harness.FanOutStreamingEngineWorkerHarnessTest > testOnNewWorkerMetadata_correctlyRemovesStaleWindmillServers FAILED
    java.lang.AssertionError at GrpcCleanupRule.java:201
```

Not sure if/how this could be related to my PR

**Comment by shunping** [comment]
Run Java PreCommit

**Review by shunping — APPROVED** [review]
LGTM. Thanks!

**Review by damccorm — APPROVED** [review]

**Pull request merged**
